src/runtime-rs/crates/hypervisor/src/remote/inner.rs

```diff
@@ -0,0 +1,325 @@
+    /// netns path
+    pub(crate) netns: Option<String>,
+
+    exit_notify: Option<mpsc::Sender<i32>>,
+    exit_waiter: Mutex<(mpsc::Receiver<i32>, i32)>,
+}
+
+impl RemoteInner {
+    pub fn new() -> Self {
+        let (exit_notify, exit_waiter) = mpsc::channel(1);
+
+        Self {
+            id: "".to_string(),
+            config: HypervisorConfig::default(),
+            agent_socket_path: "".to_string(),
+            annotations: HashMap::new(),
+            netns: None,
+
+            exit_notify: Some(exit_notify),
+            exit_waiter: Mutex::new((exit_waiter, 0)),
+        }
+    }
+
+    fn new_ttrpc_client(&self) -> Result<HypervisorClient> {
+        let c = Client::connect(&format!(
+            "unix://{}",
+            &self.config.remote_info.hypervisor_socket
+        ))?;
+        Ok(HypervisorClient::new(c))
+    }
+
+    pub(crate) async fn prepare_vm(&mut self, id: &str, netns: Option<String>) -> Result<()> {
+        info!(sl!(), "Preparing REMOTE VM");
+        self.id = id.to_string();
+
+        if let Some(netns_path) = &netns {
+            debug!(sl!(), "set netns for vmm master {:?}", &netns_path);
+            std::fs::metadata(netns_path).context("check netns path")?;
+        }
+
+        let client = self.new_ttrpc_client()?;
+
+        let ctx = context::Context::default();
+        let req = CreateVMRequest {
+            id: id.to_string(),
+            annotations: self.annotations.clone(),
+            networkNamespacePath: netns.clone().unwrap_or_default(),
+            ..Default::default()
+        };
+        info!(sl!(), "Preparing REMOTE VM req: {:?}", req.clone());
+        let resp = client.create_vm(ctx, &req).await?;
+        info!(sl!(), "Preparing REMOTE VM resp: {:?}", resp.clone());
+        self.agent_socket_path = resp.agentSocketPath;
+        self.netns = netns;
+        Ok(())
+    }
+
+    pub(crate) async fn start_vm(&mut self, timeout: i32) -> Result<()> {
+        info!(sl!(), "Starting REMOTE VM");
+
+        let mut min_timeout = DEFAULT_MIN_TIMEOUT;
+        if self.config.remote_info.hypervisor_timeout > 0 {
+            min_timeout = self.config.remote_info.hypervisor_timeout.min(timeout);
+        }
+        let timeout = min_timeout;
+
+        let client = self.new_ttrpc_client()?;
+
+        let req = StartVMRequest {
+            id: self.id.clone(),
+            ..Default::default()
+        };
+        let ctx =
+            context::with_timeout(time::Duration::from_secs(timeout as u64).as_nanos() as i64);
+        let _resp = client.start_vm(ctx, &req).await?;
+
+        let paths = fs::read_dir(Path::new(&self.agent_socket_path).parent().unwrap()).unwrap();
+        for path in paths {
+            info!(sl!(), "agent dir Name: {}", path.unwrap().path().display())
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn stop_vm(&mut self) -> Result<()> {
+        info!(sl!(), "Stopping REMOTE VM");
+
+        let client = self.new_ttrpc_client()?;
```

Comment:
Another `new_ttrpc_client()`?  Is there something that I'm missing about this call?

Reply:
You mean maybe I can make the `client` as a field inside the struct `RemoteInner`?

Reply:
Ah okay, I assumed that we have that already and overlooked that we don't - sorry, my bad!

With that being the case my point is rather moot, though I do wonder a bit what the advantage is of making a new connection every time we need to communicate.

Reply:
I just inspired by the func `openRemoteService` in golang, which opens a new connection everytime. But actually every rpc method in the ttrpc-client only takes `&self`, maybe it's a good idea to move the client to the `RemoteInner` to reduce the cost of making a new connection.

Reply:
I'm aware that the go code works like this, too.

I mean, I don't have a strong opinion at this point.  If the hypervisor protocol is a datagram one then the difference between storing a client instance vs. recreating it every time probably doesn't matter at all.  If the protocol is a stream one then it'd probably be cleaner to store the client in `RemoteInner`.

Personally I'd probably prefer storing the client if there's no strong argument against, just because that's how it's usually done.  That's just personal preference though, feel free to ignore it.

Reply:
Yes, I also think what you said is absolutely correct! I'll try to make some modifications.

Reply:
I renamed the method `new_ttrpc_client()` to `get_ttrpc_client()`, ensuring that the client is lazily created only once. Additionally, I added a field `Option<Client>` to `RemoteInner`, which requires a manual implementation of `Default` trait.